proto/asn1-pdu-proto/asn1_proto_to_der.cc

```diff
@@ -0,0 +1,158 @@
+
+uint8_t ASN1ProtoToDER::GetVariableIntLen(const uint64_t value,
+                                          const size_t base) {
+  uint8_t base_bits = log2(base);
+  for (uint8_t num_bits = (sizeof(value) - 1) * 8; num_bits >= base_bits;
+       num_bits -= base_bits) {
+    if ((value >> num_bits)) {
+      return ceil((double)num_bits / base_bits) + 1;
+    }
+  }
+  // Special-case: zero requires one, not zero bytes.
+  return 1;
+}
+
+void ASN1ProtoToDER::InsertVariableInt(const size_t value, const size_t pos) {
+  std::vector<uint8_t> variable_int;
+  for (uint8_t shift = GetVariableIntLen(value, 256); shift != 0; --shift) {
+    variable_int.push_back((value >> ((shift - 1) * 8)) & 0xFF);
+  }
+  encoder_.insert(encoder_.begin() + pos, variable_int.begin(),
+                  variable_int.end());
+}
+
+void ASN1ProtoToDER::EncodeOverrideLength(const std::string raw_len,
+                                          const size_t len_pos) {
+  encoder_.insert(encoder_.begin() + len_pos, raw_len.begin(), raw_len.end());
+}
+
+void ASN1ProtoToDER::EncodeIndefiniteLength(const size_t len_pos) {
+  encoder_.insert(encoder_.begin() + len_pos, 0x80);
+  // The PDU's value is from |len_pos| to the end of |encoder_|, so just add an
+  // EOC marker to the end.
+  encoder_.push_back(0x00);
+  encoder_.push_back(0x00);
+}
+
+void ASN1ProtoToDER::EncodeDefiniteLength(const size_t actual_len,
+                                          const size_t len_pos) {
+  InsertVariableInt(actual_len, len_pos);
+  size_t len_num_bytes = GetVariableIntLen(actual_len, 256);
+  // X.690 (2015), 8.1.3.3: The long-form is used when the length is
+  // larger than 127.
+  // Note: |len_num_bytes| is not checked here, because it will equal
+  // 1 for values [128..255], but those require the long-form length.
+  if (actual_len > 127) {
+    // See X.690 (2015) 8.1.3.5.
+    // Long-form length is encoded as a byte with the high-bit set to indicate
+    // the long-form, while the remaining bits indicate how many bytes are used
+    // to encode the length.
+    encoder_.insert(encoder_.begin() + len_pos, (0x80 | len_num_bytes));
+  }
+}
+
+void ASN1ProtoToDER::EncodeLength(const Length& len,
+                                  const size_t actual_len,
+                                  const size_t len_pos) {
+  if (len.has_length_override()) {
+    EncodeOverrideLength(len.length_override(), len_pos);
+  } else if (len.has_indefinite_form() && len.indefinite_form()) {
+    EncodeIndefiniteLength(len_pos);
+  } else {
+    EncodeDefiniteLength(actual_len, len_pos);
+  }
+}
+
+void ASN1ProtoToDER::EncodeValue(const Value& val) {
+  for (const auto& val_ele : val.val_array()) {
+    if (!kErrRecursionLimitReached) {
+      if (val_ele.has_pdu()) {
+        EncodePDU(val_ele.pdu());
+      } else {
+        encoder_.insert(encoder_.end(), val_ele.val_bits().begin(),
+                        val_ele.val_bits().end());
+      }
+    }
```

Comment:
```suggestion
    if (kErrRecursionLimitReached) {
      // If the message exceeds the recursion limit, abort processing the
      // protobuf in order to limit uninteresting work.
      return;
    }
    if (val_ele.has_pdu()) {
      EncodePDU(val_ele.pdu());
    } else {
      encoder_.insert(encoder_.end(), val_ele.val_bits().begin(),
                      val_ele.val_bits().end());
    }
```

In general, try to handle errors early and avoid introducing the horizontal depth. Each nesting level is another piece of data a reader has to keep on their mental stack, to try to remember what the conditions are for the current statement. Particularly with `if` statements, the reader has to read all the way to end of the closing brace to determine whether there is an `else` statement that might processing things differently.

By using early returns, you make it a little easier to reason about the state. Particularly for error handling, the idea is "handle the exceptional situation as appropriate, as early as possible", such as returning an error code or taking some corrective action, so that the reader can continue thinking about what "normal" processing looks like.

Reply:
Very insightful and accurate -- change accepted!